### PR TITLE
Jane's Hotel Project

### PR DESCRIPTION
**Pull request by JaneEdwMcN**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |  I chose to make a main class (hotel) that would depend on the other classes. I didn’t really consider any other options. I had no idea how to implement this project any other way. I wanted to get started, so that helped me finalize my decision. That and not coming up with any other ideas. In the future, I would like to spend more time thinking about design because I feel that this is important.
Describe a concept that you gained more clarity on as you worked on this assignment. | I think I learned more about user stories for this assignment. Also, Reading Chapter 3 of POODR enabled me to update my refactors.txt with concepts from that chapter. I’m still not 100% clear about how OOD works and how to implement it, but this gave me slightly more clarity.
Describe a _nominal test_ that you wrote for this assignment. | I tested to see if the first available room gets reserved and if the second available room gets reserved.
Describe an _edge case test_ that you wrote for this assignment. | I tested every combination for unavailable and available dates for reserving a room.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I don’t think I wrote enough pseudocode. I’m still not quite sure how to do that unfortunately. Writing the tests first helped though because it gave me a clearer image of the code I wanted to write.

### CODE REVIEW

File: lib/hotel.rb [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+class Hotel
+  attr_reader :reservations, :rooms, :room_blocks
+
+  def initialize(num)
+    @rooms = load_rooms(num)
+    @reservations = []
+    @room_blocks = []
+  end
+
+  def load_rooms(number)
```
Comment (dHelmgren) [inline lib/hotel.rb:16]: I'm of the opinion that this should be a factory method living inside your `Room` class.

File: lib/hotel.rb [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+  def list_rooms
+    room_list = []
+    rooms.each do |room|
+      room_list << "Room number: #{room.room_number}"
+    end
+    return room_list
+  end
+
+  def reserve_room(start_date, end_date, number_of_rooms)
```
Comment (dHelmgren) [inline lib/hotel.rb]: FWIW, this is the easiest to read code I've seen for reserving a room so far. great work!

File: lib/hotel.rb [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+  def dates_overlap?(start_a, end_a, start_b, end_b)
+    (Date.parse(start_a) <= end_b) && (Date.parse(end_a) > start_b)
+  end
+
+  def list_reservations(date)
+    list = []
+    @reservations.each do |reservation|
+      if reservation.start_date == Date.parse(date)
+        list << "ID: #{reservation.id}\nDates: #{reservation.start_date} - #{reservation.end_date}\nNumber of rooms: #{reservation.number_of_rooms}\nCost: #{reservation.cost}"
```
Comment (dHelmgren) [inline lib/hotel.rb]: This seems like a poor separation of concerns. Let whatever part of the program is calling this method deal with the text formatting if it's needed. This method is much less flexible for having all this string business in it.

File: lib/hotel.rb [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    raise ArgumentError, "No such reservation."
+  end
+
+  def list_available_rooms(start_d, end_d)
+    rooms_without_reservations = []
+    available_room_list = []
+
+    if rooms_not_available(start_d, end_d).nil?
+      return list_rooms
+    elsif rooms_not_available(start_d, end_d).length == rooms.length
```
Comment (dHelmgren) [inline lib/hotel.rb]: So, you run this method twice. It's fine, but you'd save your program some cycles by running it once and just checking the return value. I like this solution tho.

File: lib/hotel.rb [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+      return "All rooms are reserved from #{start_d} to #{end_d}."
+    else
+      rooms.each do |room|
+        if !(rooms_not_available(start_d, end_d).include? room)
```
Comment (dHelmgren) [inline lib/hotel.rb]: Ah shoot! You run it again here! Each time you write rooms_not_available, you run through that whole list of rooms!

File: spec/hotel_spec.rb [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+describe "Hotel Class: Wave One: Tracking Reservations" do
+  describe "hotel class initialize" do
+    before do
+      @hotel = Hotel.new(20)
+    end
+    it "creates 20 hotel rooms" do
+      expect(@hotel.rooms.length).must_equal 20
+    end
+    it "each hotel room is an instance of a room" do
+      expect(@hotel.rooms[0]).must_be_instance_of Room
```
Comment (dHelmgren) [inline spec/hotel_spec.rb]: This only checks one, why not stick it in a loop? No skin off your nose right?

File: spec/hotel_spec.rb [hunk 7]
```diff
@@ -0,0 +1,230 @@
+require_relative 'spec_helper'
+require 'pry'
+
+describe "Hotel Class: Wave One: Tracking Reservations" do
+  describe "hotel class initialize" do
+    before do
+      @hotel = Hotel.new(20)
+    end
```
Comment (dHelmgren) [inline spec/hotel_spec.rb:8]: You do this `before` for each set of tests. Did you know you can put it in the outer `describe` block and it will be run before each inner `describe` block?

File: spec/hotel_spec.rb [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+      @hotel.reserve_room('2018-03-01', '2018-03-15', 20)
+      expect{@hotel.reserve_room('2018-03-15', '2018-03-20', 1)}.must_be_silent
+    end
+
+  end
+
+  describe "list_reservations" do
+    before do
+      @hotel = Hotel.new(20)
+      @hotel.reserve_room('2018-02-03', '2018-02-05', 1)
```
Comment (dHelmgren) [inline spec/hotel_spec.rb]: What about when you give the program bad dates?

File: spec/hotel_spec.rb [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+describe "Hotel Class: Wave Two: Room Availability" do
+
+  describe "list_available_rooms" do
+    before do
+      @hotel = Hotel.new(20)
+      @hotel.reserve_room('2018-02-03', '2018-02-05', 1)
+    end
+
+    it "lists all rooms not reserved" do
+      expect(@hotel.list_available_rooms('2018-01-03', '2018-01-05').length).must_equal 20
```
Comment (dHelmgren) [inline spec/hotel_spec.rb]: What if the user hands its something that isn't a date?

**Comment by dHelmgren** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| A lot of your early check ins are REALLY big. Scope down with each checkin and add things on from one stable (not necessarily functional) piece to the next.
Answer comprehension questions	| Pseudocode can be as simple as naming methods and writing what you want them to do. Don't stress it too much, it's mostly a tool for thinking about how you will build things.
**Design** |
Each class is responsible for a single piece of the program | Generally speaking, yes!
Classes are loosely coupled | Not really. While there is little interdependency, you hit the nail on the head with your own analysis; `Hotel` is broken if it doesn't have the other classes supporting it. One approach to fix this might be by using a factory method to generate rooms, and by using keyword arguments so that you aren't dependent on the positional arguments of your methods.
**Wave 1** |
List rooms | Hotel#list_rooms
Reserve a room for a given date range | Hotel#reserve_room
List reservations for a given date | Hotel#list_reservations
Calculate reservation price | Hotel#reservation_cost
Invalid date range produces an error | Sure does! And I didn't even have to go hunting.
**Wave 2** |
View available rooms for a given date range | Hotel#list_available_rooms I like the logic, but you call one method like... a bazillion times when you only need to call it once. Food for thought!
Reserving a room that is not available produces an error | You prevent this problem from occurring.
Test coverage | Your tests need to hit more edge cases! Think like a total jerk. What would a jerk do to intentionally break your code? Test THAT.
**Wave 3** |
Create a block of rooms | Hotel#create_room_block
Check if a block has rooms | Hotel#list_available_block_rooms
Reserve a room from a block | Hotel#reserve_block_room
Test coverage | I don't see a spec file for the block class on its own!
**Fundamentals** | 
Names variables, classes and modules appropriately | Yeah!
Understanding of variable scope - local vs instance | Yes.
Can create complex logical structures utilizing variables | Yes!
Appropriately uses methods to break down tasks into smaller simpler tasks | You are a master of this. I find your code eminently readable, and I appreciate the heck out of that.
Understands the differences between class and instance methods | Yes! 
Appropriately uses iterators and enumerables | Totally!
Appropriately writes and utilizes classes | Your classes vary in how effective they are for tackling your problems, but we already plan to talk about this.
Appropriately utilizes modules as a mechanism to encapsulate functionality | Yes!
**Wrap Up** |
There is a refactors.txt file | Yes there is!
The file provides a roadmap to future changes | AND IT NAILS THE THINGS I SAID HERE TOO! NOICE!
**Additional Feedback** | Okay, so you do this thing with `rooms_not_available` that is a bad habit, and you ought to break it. When a method returns a value that you want to check more than once in the same scope (say, inside a single method) do the method once, and assign the return value to a variable! Otherwise, that code is getting run over and over again. `rooms_not_available` is particularly thorny because it has loops inside of itself, and so when it gets called as part of a loop, the loop triggers more loops inside the original loop! Not efficient! All that said, I think this is really clear and solid code, barring a couple of small missteps!
